src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,150 @@
+
+For example, consider a table as showed in the image again:
+
+![Figure 5](percolator-figure-3.png)
+
+Let's read Bob's balance.
+1. Get a timestamp. Assume we got `9`.
+2. Check lock. The row of Bob is not locked, so we continue.
+3. Get the latest record in `write` column that committed before `9`. We got the record with `commit_ts` equals to `8`, and it says the `start_ts` is `7`, which means, it's corresponding data is at timestamp `7` in `data` column.
+4. Get the value in `data` column whose timestamp is exactly `7`. We got `$3`.
+
+This algorithm provides us both lock-free reading and the ability to perform historical read. In the above example, if we specify that we want to read at time point `7`, then we will se the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.
+
+### Handling Conflicts
+
+Conflicts are identified by checking the `lock` column. A row can have many versions of data, but it can have at most one lock at any time.
+
+When we are performing a write operation, we tries to lock every affected rows in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in a scenario that conflicts occur frequently.
+
+To roll back a row, just simply remove its lock and its corresponding value in `data` column.
+
+### Tolerating crashes
+
+Percolator has the ability to survive crashes without breaking correctness of data.
+
+First, let's see what will happen after a crash. A crash may happen during `Prewrite`, during `Commit` or between these two. But any way, we can simply divide these conditions into two types: before committing the primary, or after committing the primary.
+
+So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it in fact doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.
+
+* If the primary lock disappeared but there's a record `data @ T0.start_ts` in `write` column, it means that `T0` has been successfully committed. Then the row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After that, the new transaction `T1` can continue.
+* If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.
+* If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.
+* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback it self, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
```

Comment:
```suggestion
* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
```